File: pkg-r/tests/testthat/test-handoff_view.R
new_handoff_fake_session <- function() {
  session <- new.env(parent = emptyenv())
  session$messages <- list()
  session$ns <- function(id) paste0("module-", id)
  session$sendCustomMessage <- function(type, message) {
    session$messages[[length(session$messages) + 1L]] <- list(
      type = type,
      payload = message
    )
  }
  session
}

new_handoff_fake_chat <- function() {
  chat <- new.env(parent = emptyenv())
  chat$appended <- list()
  chat$append <- function(response, role = "assistant", icon = NULL) {
    chat$appended[[length(chat$appended) + 1L]] <- list(
      response = response,
      role = role,
      icon = icon
    )
  }
  chat
}

new_handoff_view <- function() {
  session <- new_handoff_fake_session()
  chat_module <- new_handoff_fake_chat()
  list(
    view = HandoffView$new(
      session = session,
      chat_module = chat_module
    ),
    session = session,
    chat_module = chat_module
  )
}

describe("HandoffView$new()", {
  it("keeps dependencies and derived IDs private", {
    fixture <- new_handoff_view()
    private_names <- c(
      "session",
      "chat_module",
      "panel_root_id",
      "modal_root_id",
      "editor_id",
      "directions_id",
      "open_input_id"
    )

    expect_length(intersect(names(fixture$view), private_names), 0L)
  })
})

describe("HandoffView$set_panel_open()", {
  it("sends panel toggle messages", {
    fixture <- new_handoff_view()
    view <- fixture$view

    view$set_panel_open(TRUE)
    view$set_panel_open(FALSE)

    expect_identical(
      fixture$session$messages,
      list(
        list(
          type = "querychat-handoff-panel-toggle",
          payload = list(
            root_id = "module-handoff_root",
            open = TRUE
          )
        ),
        list(
          type = "querychat-handoff-panel-toggle",
          payload = list(
            root_id = "module-handoff_root",
            open = FALSE
          )
        )
      )
    )
  })
})

describe("HandoffView$clear_source()", {
  it("clears the editor and disables downloads", {
    fixture <- new_handoff_view()
    view <- fixture$view

    view$clear_source("r")

    expect_identical(
      fixture$session$messages,
      list(
        list(
          type = "querychat-handoff-source-update",
          payload = list(
            root_id = "module-handoff_root",
            id = "module-handoff_source_editor",
            value = "",
            language = "r",
            download_available = FALSE
          )
        )
      )
    )
  })
})

describe("HandoffView$replace_source()", {
  it("replaces the editor source", {
    fixture <- new_handoff_view()
    view <- fixture$view

    view$replace_source("print(1)")

    expect_identical(
      fixture$session$messages[[1]],
      list(
        type = "querychat-handoff-source-update",
        payload = list(
          root_id = "module-handoff_root",
          id = "module-handoff_source_editor",
          value = "print(1)"
        )
      )
    )
  })
})

describe("HandoffView$append_source()", {
  it("appends an editor source delta", {
    fixture <- new_handoff_view()
    view <- fixture$view

    view$append_source("\nprint(2)")

    expect_identical(
      fixture$session$messages[[1]],
      list(
        type = "querychat-handoff-source-update",
        payload = list(
          root_id = "module-handoff_root",
          id = "module-handoff_source_editor",
          value = "\nprint(2)",
          append = TRUE
        )
      )
    )
  })
})

describe("HandoffView$set_streaming()", {
  it("sends streaming state changes", {
    fixture <- new_handoff_view()
    view <- fixture$view

    view$set_streaming(TRUE)
    view$set_streaming(FALSE)

    expect_identical(
      fixture$session$messages,
      list(
        list(
          type = "querychat-handoff-streaming",
          payload = list(
            root_id = "module-handoff_root",
            active = TRUE
          )
        ),
        list(
          type = "querychat-handoff-streaming",
          payload = list(
            root_id = "module-handoff_root",
            active = FALSE
          )
        )
      )
    )
  })
})

describe("HandoffView$show_handoff()", {
  it("shows the current source, language, and download state", {
    fixture <- new_handoff_view()
    view <- fixture$view
    state <- HandoffState(
      handoff_id = "handoff-123",
      handoff_type = resolve_handoff_type("quarto-dashboard", "r"),
      system_prompt = "System prompt",
      source = "library(shiny)"
    )

    view$show_handoff(state, download_available = FALSE)

    expect_identical(
      fixture$session$messages[[1]],
      list(
        type = "querychat-handoff-source-update",
        payload = list(
          root_id = "module-handoff_root",
          id = "module-handoff_source_editor",
          value = "library(shiny)",
          language = "markdown",
          download_available = FALSE
        )
      )
    )
  })
})

describe("HandoffView$show_recommendation()", {
  it("shows a deduplicated recommendation in the modal", {
    fixture <- new_handoff_view()
    view <- fixture$view
    recommendation <- HandoffRecommendation(
      selected_ids = c("query-0", "viz-1", "query-0"),
      format_id = "shiny-app",
      directions = "Use a sidebar."
    )

    view$show_recommendation(recommendation)

    expect_identical(
      fixture$session$messages[[1]],
      list(
        type = "querychat-handoff-recommend",
        payload = list(
          root_id = "module-handoff_modal_root",
          selected_ids = list("query-0", "viz-1"),
          format_id = "shiny-app",
          directions = "Use a sidebar.",
          directions_id = "module-handoff_directions"
        )
      )
    )
  })
})

describe("HandoffView$show_recommendation_error()", {
  it("shows the recommendation error in the modal", {
    fixture <- new_handoff_view()
    view <- fixture$view

    view$show_recommendation_error("Service unavailable")

    expect_identical(
      fixture$session$messages[[1]],
      list(
        type = "querychat-handoff-recommend-error",
        payload = list(
          root_id = "module-handoff_modal_root",
          error = "Service unavailable"
        )
      )
    )
  })
})

describe("HandoffView$show_modal()", {
  it("shows the namespaced modal in its session", {
    shown <- NULL
    local_mocked_bindings(
      showModal = function(ui, session) {
        shown <<- list(ui = ui, session = session)
      },
      .package = "shiny"
    )
    fixture <- new_handoff_view()
    view <- fixture$view

    view$show_modal(list())

    expect_identical(shown$session, fixture$session)
    expect_match(
      as.character(shown$ui),
      'id="module-handoff_modal_root"',
      fixed = TRUE
    )
  })
})

describe("HandoffView$remove_modal()", {
  it("removes the modal from its session", {
    removed_from <- NULL
    local_mocked_bindings(
      removeModal = function(session) {
        removed_from <<- session
      },
      .package = "shiny"
    )
    fixture <- new_handoff_view()
    view <- fixture$view

    view$remove_modal()

    expect_identical(removed_from, fixture$session)
  })
})

describe("HandoffView$append_pill()", {
  it("appends the pill and rendered summary as one assistant message", {
    fixture <- new_handoff_view()
    view <- fixture$view
    handoff_type <- resolve_handoff_type("quarto-dashboard", "python")

    view$append_pill("handoff-123", handoff_type, "A **dashboard**")

    expect_length(fixture$chat_module$appended, 1L)
    appended <- fixture$chat_module$appended[[1]]
    markup <- as.character(appended$response)
    expect_identical(appended$role, "assistant")
    expect_null(appended$icon)
    expect_match(
      markup,
      'data-handoff-id="handoff-123"',
      fixed = TRUE
    )
    expect_match(markup, "<p>A **dashboard**</p>", fixed = TRUE)
    expect_no_match(markup, "<strong>", fixed = TRUE)
  })

  it("escapes untrusted summary markup before appending it", {
    fixture <- new_handoff_view()
    view <- fixture$view
    handoff_type <- resolve_handoff_type("quarto-dashboard", "python")
    summary <- paste0(
      '<img src=x onerror="alert(1)"> ',
      "[click](javascript:alert(2)) ",
      "<script>alert(3)</script>"
    )

    view$append_pill("handoff-123", handoff_type, summary)

    markup <- as.character(fixture$chat_module$appended[[1]]$response)
    expect_no_match(markup, "<img", fixed = TRUE)
    expect_no_match(markup, "<a href=", fixed = TRUE)
    expect_no_match(markup, "<script", fixed = TRUE)
    expect_match(
      markup,
      '&lt;img src=x onerror="alert(1)"&gt;',
      fixed = TRUE
    )
    expect_match(
      markup,
      "[click](javascript:alert(2))",
      fixed = TRUE
    )
    expect_match(
      markup,
      "&lt;script&gt;alert(3)&lt;/script&gt;",
      fixed = TRUE
    )
  })

  it("omits an empty summary", {
    fixture <- new_handoff_view()
    view <- fixture$view
    handoff_type <- resolve_handoff_type("quarto-dashboard", "python")

    view$append_pill("handoff-123", handoff_type, "")

    markup <- as.character(fixture$chat_module$appended[[1]]$response)
    expect_match(
      markup,
      'data-handoff-id="handoff-123"',
      fixed = TRUE
    )
    expect_no_match(markup, "<p>", fixed = TRUE)
  })
})

File: pkg-py/tests/test_handoff_view.py
import asyncio

import pytest
from pydantic import ValidationError
from querychat._handoff_protocol import SourceUpdateMessage
from querychat._handoff_state import HandoffState
from querychat._handoff_types import resolve_handoff_type
from querychat._handoff_view import HandoffView


def test_source_update_message_uses_protocol_action_and_payload():
    message = SourceUpdateMessage(
        root_id="handoff_root",
        id="handoff_source_editor",
        value="print(1)",
    )

    assert message.message_type() == "querychat-handoff-source-update"
    assert message.payload() == {
        "root_id": "handoff_root",
        "id": "handoff_source_editor",
        "value": "print(1)",
    }


def test_protocol_messages_reject_unknown_payload_fields():
    with pytest.raises(ValidationError, match="extra_field"):
        SourceUpdateMessage(
            root_id="handoff_root",
            id="handoff_source_editor",
            value="print(1)",
            extra_field=True,
        )


class FakeSession:
    def __init__(self):
        self.messages = []

    def ns(self, name):
        return f"ns-{name}"

    async def send_custom_message(self, msg_type, payload):
        self.messages.append((msg_type, payload))


class FakeChatUI:
    def __init__(self):
        self.appended = []
        self.streamed = []

    async def append_message(self, message):
        self.appended.append(message)

    async def append_message_stream(self, stream):
        async for part in stream:
            self.streamed.append(part)


def make_view():
    return HandoffView(FakeSession(), FakeChatUI())


class TestUpdateSource:
    def test_sends_source_update_to_editor(self):
        view = make_view()
        asyncio.run(view.update_source("print(1)"))
        assert view.session.messages == [
            (
                "querychat-handoff-source-update",
                {
                    "root_id": view.panel_root_id,
                    "id": view.editor_id,
                    "value": "print(1)",
                },
            ),
        ]

    def test_appends_source_delta_to_editor(self):
        view = make_view()
        asyncio.run(view.append_source("print(1)"))

        assert view.session.messages == [
            (
                "querychat-handoff-source-update",
                {
                    "root_id": view.panel_root_id,
                    "id": view.editor_id,
                    "value": "print(1)",
                    "append": True,
                },
            ),
        ]

    def test_show_handoff_sends_current_source_and_download_state(self):
        view = make_view()
        handoff_type = resolve_handoff_type("quarto-dashboard", "python")
        state = HandoffState(
            handoff_id="a",
            handoff_type=handoff_type,
            system_prompt="sys",
            source="print(1)",
        )

        asyncio.run(view.show_handoff(state, download_available=False))

        assert view.session.messages == [
            (
                "querychat-handoff-source-update",
                {
                    "root_id": view.panel_root_id,
                    "id": view.editor_id,
                    "value": "print(1)",
                    "language": handoff_type.editor_language,
                    "download_available": False,
                },
            ),
        ]


class TestSetStreaming:
    def test_toggles_streaming_flag(self):
        view = make_view()
        asyncio.run(view.set_streaming(active=True))
        asyncio.run(view.set_streaming(active=False))
        assert view.session.messages == [
            (
                "querychat-handoff-streaming",
                {"root_id": view.panel_root_id, "active": True},
            ),
            (
                "querychat-handoff-streaming",
                {"root_id": view.panel_root_id, "active": False},
            ),
        ]


class TestAppendPill:
    def test_appends_complete_pill_message_with_summary(self):
        view = make_view()
        handoff_type = resolve_handoff_type("quarto-dashboard", "python")
        asyncio.run(view.append_pill("abc123", handoff_type, "A dashboard"))

        assert len(view.chat_ui.appended) == 1
        message = str(view.chat_ui.appended[0])
        assert "abc123" in message
        assert "<p>A dashboard</p>" in message
        assert view.chat_ui.streamed == []

    def test_omits_empty_summary(self):
        view = make_view()
        handoff_type = resolve_handoff_type("quarto-dashboard", "python")
        asyncio.run(view.append_pill("abc123", handoff_type, ""))

        assert len(view.chat_ui.appended) == 1
        assert "abc123" in str(view.chat_ui.appended[0])
        assert view.chat_ui.streamed == []


class FakeUI:
    def __init__(self):
        self.shown = []
        self.removed = 0

    def modal_show(self, modal):
        self.shown.append(modal)

    def modal_remove(self):
        self.removed += 1


class TestModal:
    def test_show_modal_delegates_to_ui(self, monkeypatch):
        fake_ui = FakeUI()
        monkeypatch.setattr("querychat._handoff_view.ui", fake_ui)
        monkeypatch.setattr(
            "querychat._handoff_view.build_modal_ui",
            lambda ns, items: "MODAL",
        )
        view = make_view()
        view.show_modal([])
        assert fake_ui.shown == ["MODAL"]

    def test_remove_modal_delegates_to_ui(self, monkeypatch):
        fake_ui = FakeUI()
        monkeypatch.setattr("querychat._handoff_view.ui", fake_ui)
        view = make_view()
        view.remove_modal()
        assert fake_ui.removed == 1
